File: benchmarks/suite/binary_trees/bench.cpp
#include <cstdio>

struct TreeNode {
    TreeNode* left;
    TreeNode* right;
};

TreeNode* make_tree(int depth) {
    TreeNode* node = new TreeNode();
    if (depth == 0) {
        node->left = nullptr;
        node->right = nullptr;
    } else {
        node->left = make_tree(depth - 1);
        node->right = make_tree(depth - 1);
    }
    return node;
}

int check_tree(TreeNode* node) {
    if (node->left == nullptr) return 1;
    return 1 + check_tree(node->left) + check_tree(node->right);
}

void free_tree(TreeNode* node) {
    if (node->left) {
        free_tree(node->left);
        free_tree(node->right);
    }
    delete node;
}

int main() {
    int min_depth = 4;
    int max_depth = 10;
    int stretch_depth = max_depth + 1;

    TreeNode* stretch = make_tree(stretch_depth);
    printf("stretch tree of depth %d\t check: %d\n", stretch_depth, check_tree(stretch));
    free_tree(stretch);

    TreeNode* long_lived = make_tree(max_depth);

    for (int d = min_depth; d <= max_depth; d += 2) {
        int iterations = 1;
        for (int i = 0; i < max_depth - d + min_depth; i++) {
            iterations *= 2;
        }

        int check = 0;
        for (int i = 1; i <= iterations; i++) {
            TreeNode* t = make_tree(d);
            check += check_tree(t);
            free_tree(t);
        }

        printf("%d\t trees of depth %d\t check: %d\n", iterations, d, check);
    }

    printf("long lived tree of depth %d\t check: %d\n", max_depth, check_tree(long_lived));
    free_tree(long_lived);
    return 0;
}

File: benchmarks/suite/binary_trees/bench.py
"""OBO Benchmark: Binary Trees (Benchmarks Game)
Tests: recursive object allocation, tree traversal, memory management"""

import sys
sys.setrecursionlimit(1000000)


class TreeNode:
    __slots__ = ('left', 'right')
    def __init__(self, left, right):
        self.left = left
        self.right = right


def make_tree(depth):
    if depth == 0:
        return TreeNode(None, None)
    return TreeNode(make_tree(depth - 1), make_tree(depth - 1))


def check_tree(node):
    if node.left is None:
        return 1
    return 1 + check_tree(node.left) + check_tree(node.right)


def main():
    min_depth = 4
    max_depth = 10
    stretch_depth = max_depth + 1

    stretch = make_tree(stretch_depth)
    print("stretch tree of depth " + str(stretch_depth) + "\t check: " + str(check_tree(stretch)))

    long_lived = make_tree(max_depth)

    d = min_depth
    while d <= max_depth:
        iterations = 1
        for _ in range(max_depth - d + min_depth):
            iterations *= 2

        check = 0
        for _ in range(1, iterations + 1):
            check += check_tree(make_tree(d))

        print(str(iterations) + "\t trees of depth " + str(d) + "\t check: " + str(check))
        d += 2

    print("long lived tree of depth " + str(max_depth) + "\t check: " + str(check_tree(long_lived)))


main()
